src/UnityExtension/Assets/Editor/GitHub.Unity/UI/TreeControl.cs

```diff
@@ -0,0 +1,465 @@
+                }
+
+                if (node.Level < level)
+                {
+                    for (; node.Level > level && indents.Count > 1; level--)
+                    {
+                        Unindent();
+                    }
+                }
+                level = node.Level;
+
+                if (!node.IsHidden)
+                {
+                    if (visible)
+                    {
+                        RequiresRepaint = HandleInput(rect, node, i, singleClick, doubleClick);
+                    }
+                    rect.y += ItemHeight + ItemSpacing;
+                }
+            }
+
+            Unindent();
+
+            Profiler.EndSample();
+            return rect;
+        }
+
+        public void Focus()
+        {
+            bool selectionChanged = false;
+            if (Event.current.type == EventType.KeyDown)
+            {
+                int directionY = Event.current.keyCode == KeyCode.UpArrow ? -1 : Event.current.keyCode == KeyCode.DownArrow ? 1 : 0;
+                int directionX = Event.current.keyCode == KeyCode.LeftArrow ? -1 : Event.current.keyCode == KeyCode.RightArrow ? 1 : 0;
+                if (directionY != 0 || directionX != 0)
+                {
+                    if (directionY < 0 || directionY < 0)
+                    {
+                        SelectedNode = nodes[nodes.Count - 1];
+                        selectionChanged = true;
+                        Event.current.Use();
+                    }
+                    else if (directionY > 0 || directionX > 0)
+                    {
+                        SelectedNode = nodes[0];
+                        selectionChanged = true;
+                        Event.current.Use();
+                    }
+                }
+            }
+            RequiresRepaint = selectionChanged;
+        }
+
+        public void Blur()
+        {
+            SelectedNode = null;
+            RequiresRepaint = true;
+        }
+
+        private int ToggleNodeVisibility(int idx, TreeNode rootNode)
+        {
+            var rootNodeLevel = rootNode.Level;
+            rootNode.IsCollapsed = !rootNode.IsCollapsed;
+            idx++;
+            for (; idx < nodes.Count && nodes[idx].Level > rootNodeLevel; idx++)
+            {
+                nodes[idx].IsHidden = rootNode.IsCollapsed;
+                if (nodes[idx].IsFolder && !rootNode.IsCollapsed && nodes[idx].IsCollapsed)
+                {
+                    var level = nodes[idx].Level;
+                    for (idx++; idx < nodes.Count && nodes[idx].Level > level; idx++) { }
+                    idx--;
+                }
+            }
+            if (SelectedNode != null && SelectedNode.IsHidden)
+            {
+                SelectedNode = rootNode;
+            }
+            return idx;
+        }
+
+        private bool HandleInput(Rect rect, TreeNode currentNode, int index, Action<TreeNode> singleClick = null, Action<TreeNode> doubleClick = null)
```

Comment:
So are singleClick and doubleClick what gets displayed if a node gets single or double clicked?

Reply:
Almost... they are the function delegates that get called when a node gets single or double clicked

Reply:
Got it, thanks!